File: pos-maui/Modelo/Dao/ProductoDao.cs
using Dapper;
using PosMaui.Modelo.Entidad;

namespace PosMaui.Modelo.Dao;

public static class ProductoDao
{
    public static List<Producto> ObtenerTodos()
    {
        using var conn = ConexionBD.Abrir();
        return conn.Query<Producto>("""
            SELECT p.id AS Id, p.sku AS Sku, p.nombre AS Nombre,
                   COALESCE(p.marca,'') AS Marca,
                   COALESCE(c.nombre,'Sin categoria') AS Categoria,
                   COALESCE(p.categoria_id,0) AS CategoriaId,
                   p.precio AS Precio,
                   COALESCE(p.unidad,'PZA') AS Unidad,
                   COALESCE(i.stock_actual,0) AS StockActual,
                   COALESCE(i.stock_minimo,0) AS StockMinimo,
                   p.activo AS Activo,
                   p.imagen_local AS ImagenLocal
            FROM productos p
            LEFT JOIN categorias c ON c.id = p.categoria_id
            LEFT JOIN inventario i ON i.producto_id = p.id
            WHERE p.activo = true
            ORDER BY p.nombre
            """).ToList();
    }

    public static List<Producto> BuscarPorTexto(string texto)
    {
        using var conn = ConexionBD.Abrir();
        return conn.Query<Producto>("""
            SELECT p.id AS Id, p.sku AS Sku, p.nombre AS Nombre,
                   COALESCE(p.marca,'') AS Marca,
                   COALESCE(c.nombre,'Sin categoria') AS Categoria,
                   p.precio AS Precio,
                   COALESCE(p.unidad,'PZA') AS Unidad,
                   COALESCE(i.stock_actual,0) AS StockActual,
                   COALESCE(i.stock_minimo,0) AS StockMinimo,
                   p.activo AS Activo,
                   p.imagen_local AS ImagenLocal
            FROM productos p
            LEFT JOIN categorias c ON c.id = p.categoria_id
            LEFT JOIN inventario i ON i.producto_id = p.id
            WHERE p.activo = true
              AND (LOWER(p.nombre) LIKE @Texto OR LOWER(p.sku) LIKE @Texto OR LOWER(COALESCE(p.marca,'')) LIKE @Texto)
            ORDER BY p.nombre
            LIMIT 80
            """, new { Texto = $"%{texto.ToLower()}%" }).ToList();
    }

    public static List<Producto> ObtenerBajoStock()
    {
        using var conn = ConexionBD.Abrir();
        return conn.Query<Producto>("""
            SELECT p.id AS Id, p.sku AS Sku, p.nombre AS Nombre,
                   COALESCE(c.nombre,'') AS Categoria,
                   i.stock_actual AS StockActual, i.stock_minimo AS StockMinimo
            FROM productos p
            JOIN inventario i ON i.producto_id = p.id
            LEFT JOIN categorias c ON c.id = p.categoria_id
            WHERE i.stock_actual <= i.stock_minimo AND p.activo = true
            ORDER BY i.stock_actual
            """).ToList();
    }

    public static void ActualizarStock(int productoId, int cantidad, string tipo)
    {
        using var conn = ConexionBD.Abrir();
        string op = tipo == "SALIDA" ? "-" : "+";
        conn.Execute($"""
            UPDATE inventario
            SET stock_actual = stock_actual {op} @Cantidad
            WHERE producto_id = @ProductoId
            """, new { Cantidad = cantidad, ProductoId = productoId });
    }
}

File: pos-python/db/producto_dao.py
from db.connection import DB


def find_all(solo_activos=True):
    cond = "WHERE p.activo = TRUE" if solo_activos else ""
    return DB.fetchall(
        f"""SELECT p.id, p.sku, p.nombre, p.precio, p.activo,
                   c.nombre AS categoria,
                   COALESCE(i.stock_actual, 0) AS stock
            FROM productos p
            LEFT JOIN categorias c ON c.id = p.categoria_id
            LEFT JOIN inventario i ON i.producto_id = p.id
            {cond}
            ORDER BY p.nombre"""
    )


def find_by_categoria(categoria_id: int):
    return DB.fetchall(
        """SELECT p.id, p.sku, p.nombre, p.precio,
                  COALESCE(i.stock_actual, 0) AS stock
           FROM productos p
           LEFT JOIN inventario i ON i.producto_id = p.id
           WHERE p.activo = TRUE AND p.categoria_id = %s
           ORDER BY p.nombre""",
        (categoria_id,)
    )


def search(texto: str):
    q = f"%{texto}%"
    return DB.fetchall(
        """SELECT p.id, p.sku, p.nombre, p.precio,
                  COALESCE(i.stock_actual, 0) AS stock
           FROM productos p
           LEFT JOIN inventario i ON i.producto_id = p.id
           WHERE p.activo = TRUE
             AND (p.nombre ILIKE %s OR p.sku ILIKE %s)
           ORDER BY p.nombre LIMIT 50""",
        (q, q)
    )


def insert(sku, nombre, precio, categoria_id=None):
    row = DB.execute_returning(
        """INSERT INTO productos (sku, nombre, precio, categoria_id)
           VALUES (%s, %s, %s, %s) RETURNING id""",
        (sku, nombre, precio, categoria_id)
    )
    if row:
        DB.execute(
            "INSERT INTO inventario (producto_id, stock_actual) VALUES (%s, 0)",
            (row["id"],)
        )
    return row


def update(prod_id, sku, nombre, precio, categoria_id=None):
    DB.execute(
        """UPDATE productos SET sku=%s, nombre=%s, precio=%s, categoria_id=%s
           WHERE id=%s""",
        (sku, nombre, precio, categoria_id, prod_id)
    )


def toggle_activo(prod_id: int):
    DB.execute(
        "UPDATE productos SET activo = NOT activo WHERE id=%s", (prod_id,)
    )
